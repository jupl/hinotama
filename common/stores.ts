import {useStrict} from 'mobx'

type StoreBuilder<S> = (isServer: boolean) => S

export function buildStoreGetter<S>(createStore: StoreBuilder<S>) {
  let store: S
  useStrict(true)
  return function getStore(isServer: boolean) {
    if(!store) {
      store = createStore(isServer)
    }
    return store
  }
}
