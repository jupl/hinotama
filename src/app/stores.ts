import * as colors from 'colors.css'
import {Color} from '../color'

/**
 * Create stores used by application
 * @return Object containing store instances
 */
export function createStores() {
  return {
    color: new Color([
      colors.teal,
      colors.green,
      colors.yellow,
      colors.orange,
    ]),
  }
}
