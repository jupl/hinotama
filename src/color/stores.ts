import {action, computed, observable} from 'mobx'

export interface IColor {
  readonly value: string
  previousColor(): void
  nextColor(): void
}

export class Color implements IColor {
  private values: string[]
  private id = 0
  @observable private index = 0

  constructor(values: string[]) {
    this.values = values
  }

  public startAutoCycle(interval: number) {
    clearInterval(this.id)
    this.id = window.setInterval(this.autoNextColor, interval)
  }

  @computed
  get value() {
    return this.values[this.index]
  }

  @action.bound
  public previousColor() {
    clearInterval(this.id)
    this.index = (this.index + this.values.length - 1) % this.values.length
  }

  @action.bound
  public nextColor() {
    clearInterval(this.id)
    this.autoNextColor()
  }

  @action.bound
  private autoNextColor() {
    this.index = (this.index + 1) % this.values.length
  }
}
