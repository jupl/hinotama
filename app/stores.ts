import {buildStoreGetter} from '../common/stores'

export interface Stores {}

export const getStores = buildStoreGetter(createStores)

export function createStores(isServer: boolean): Readonly<Stores> {
  const stores: Stores = {}
  if(isServer) {
    // Write server specific logic
  }
  else {
    // Write client specific logic
  }
  return stores
}
