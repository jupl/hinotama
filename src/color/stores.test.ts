import {Color} from './stores'

describe('Color store', () => {
  it('should not allow an empty values list', () => {
    expect(() => new Color([])).toThrow()
    expect(() => new Color(['#ffffff'])).not.toThrow()
  })
})
