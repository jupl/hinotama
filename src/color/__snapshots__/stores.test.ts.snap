// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Color store automatic cycling should cycle color automatically 1`] = `"red"`;

exports[`Color store automatic cycling should cycle color automatically 2`] = `"green"`;

exports[`Color store automatic cycling should cycle color automatically 3`] = `"blue"`;

exports[`Color store automatic cycling should cycle color automatically 4`] = `"red"`;

exports[`Color store automatic cycling should stop color cyling when calling nextColor 1`] = `"red"`;

exports[`Color store automatic cycling should stop color cyling when calling nextColor 2`] = `"green"`;

exports[`Color store automatic cycling should stop color cyling when calling nextColor 3`] = `"blue"`;

exports[`Color store automatic cycling should stop color cyling when calling previousColor 1`] = `"red"`;

exports[`Color store automatic cycling should stop color cyling when calling previousColor 2`] = `"green"`;

exports[`Color store automatic cycling should stop color cyling when calling previousColor 3`] = `"red"`;

exports[`Color store should cycle colors with nextColor 1`] = `"red"`;

exports[`Color store should cycle colors with nextColor 2`] = `"green"`;

exports[`Color store should cycle colors with nextColor 3`] = `"blue"`;

exports[`Color store should cycle colors with nextColor 4`] = `"red"`;

exports[`Color store should cycle colors with previousColor 1`] = `"red"`;

exports[`Color store should cycle colors with previousColor 2`] = `"blue"`;

exports[`Color store should cycle colors with previousColor 3`] = `"green"`;

exports[`Color store should cycle colors with previousColor 4`] = `"red"`;
